File: tiq-e/trees/3/main.go
package main

import (
	"fmt"
	"time"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isSymmetric(root *TreeNode) bool {
	if root == nil {
		return true
	}

	return isMirror(root.Left, root.Right)
}

func isMirror(tree1 *TreeNode, tree2 *TreeNode) bool {
	if tree1 == nil && tree2 == nil {
		return true
	}
	if tree1 == nil || tree2 == nil {
		return false
	}
	if tree1.Val != tree2.Val {
		return false
	}

	return tree1.Val == tree2.Val && isMirror(tree1.Left, tree2.Right) && isMirror(tree1.Right, tree2.Left)
}

func main() {
	var root = TreeNode{
		Val:   5,
		Left:  nil,
		Right: nil,
	}

	var start = time.Now()
	var result = isSymmetric(root)
	var end = time.Since(start)
	fmt.Println("Duration: ", end)
	fmt.Println("Result: ", result)
}

File: tiq-e/trees/3/main.rs
use std::cell::RefCell;
use std::rc::Rc;
use std::time::Instant;

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

struct Solution;

impl Solution {
    pub fn is_symmetric(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
        match root {
            Some(r) => Solution::is_mirror(&r.borrow().left, &r.borrow().right),
            None => true,
        }
    }

    fn is_mirror(
        tree1: &Option<Rc<RefCell<TreeNode>>>,
        tree2: &Option<Rc<RefCell<TreeNode>>>,
    ) -> bool {
        match (tree1, tree2) {
            (None, None) => true,
            (Some(tree1), Some(tree2)) => {
                tree1.borrow().val == tree2.borrow().val
                    && Solution::is_mirror(&tree1.borrow().left, &tree2.borrow().right)
                    && Solution::is_mirror(&tree1.borrow().right, &tree2.borrow().left)
            }
            _ => false,
        }
    }
}

fn main() {
    let root = TreeNode(5, None, None);

    let start = Instant::now();
    let result = Solution::max_depth(root);
    let end = start.elapsed().as_micros();
    println!("Duration: {:?}", end);
    println!("Result: {:?}", result);
}
